### DISCUSSION TITLE
distributed trace

### DISCUSSION THREAD

**Discussion by ArcticUnicorn** [istio/istio#52020] [General]
When the istio-proxy generated the trace information, Where can I get the header after generating the trace information? I use Zipkin to collect trace information and write a lua filter, but i can't get x-b3-traceid and x-b3-spanid from request header it the first istio-proxy, i woulde like to know if the tracing information is not generated in the function envoy_on_request(request_handle)，Thanks for any help!

**Comment by zirain** [comment]
This's something need to dig into envoy, you cannot get the header before `injectContext` excuted.

**Reply by ArcticUnicorn** [reply]
Thanks. Let me confirm, does that mean injectContext is executed after lua filter  【function envoy_on_request(request_handle)】? If I want to get the header with x-b3-traceid and x-b3-spanid added, is there a way to do it?

**Reply by zirain** [reply]
can you try `envoy_on_response` ?

**Reply by ArcticUnicorn** [reply]
I have already tried, client[sleep] -> server[http_bin]

client: envoy_on_request and envoy_on_response
![client](https://github.com/user-attachments/assets/e6ca1039-ba72-4417-b27c-c52ef417c7b2)

server: envoy_on_request and envoy_on_response
![server](https://github.com/user-attachments/assets/ccdc0863-935a-478c-a013-8982b4c951e2)

It seems that the trace information is added after envoy_on_request，Because the client does not have trace information, on the contrary, the server's envoy_on_request has trace information.

**Reply by zirain** [reply]
can you try client with envoy_on_response?

**Reply by ArcticUnicorn** [reply]
the response_header result is client with envoy_on_response, I executed both envoy_on_request and envoy_on_response.

**Reply by zirain** [reply]
not sure about how to do it in lua, but I can get it with wasm:

```golang
func (ctx *httpContext) OnHttpStreamDone() {
	proxywasm.LogErrorf("OnHttpStreamDone")

	reqHeaders, err := proxywasm.GetHttpRequestHeaders()
	if err != nil {
		proxywasm.LogErrorf("failed to get request headers: %v", err)
		return
	}
	for _, h := range reqHeaders {
		proxywasm.LogErrorf("request header: <%s: %s>", h[0], h[1])
	}
	return
}
```

**Reply by ArcticUnicorn** [reply]
Could you please send me the complete process of using wasm? One more question, can I adjust the order of the message headers using your method? For example, put the traceid, spanid and other information at the front of the header? Thanks

**Reply by zirain** [reply]
https://github.com/zirain/playground/tree/master/go-playgroud/tools/wasm-extensions/display-metadata

**Reply by ArcticUnicorn** [reply]
Thanks. I haven't used wasm before. I will learn it and verify whether the order of headers can be adjusted.
